### PR TITLE
Update tests and README.md to support  Juicy/juicy-html@2.0.0

### PR DESCRIPTION
**Pull request by tomalec**

### CODE REVIEW

File: test/juicy-html_content-legacy/no-content.html [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                errorSpy = sinon.spy(console, 'error');
+                errorSpy.displayName = 'console.error';
+            });
+            afterEach(function () {
+                infoSpy.restore();
+                warnSpy.restore();
+                errorSpy.restore();
+            });
+            before(function () {
+                var server = sinon.fakeServer.create();
```
Comment (warpech) [inline test/juicy-html_content-legacy/no-content.html]: Sinon advises to restore the original XHR by calling `server.restore();` after the tests.

Every test suite that uses `fakeServer` should restore the fake server after the tests.

See http://sinonjs.org/releases/v2.1.0/fake-xhr-and-server/

File: test/juicy-html_no-content.html [hunk 2]
```diff
@@ -59,15 +59,18 @@
                     //whenever the this returns true the request will not faked
                     return !url.match(/mock\//);
                 });
-
+                server.respondWith('GET', '/mock/204', [204, {"Content-Type": "text/html"}, '']);
+                server.respondWith('GET', '/mock/200', [200, {"Content-Type": "text/html"}, '']);
+                server.respondWith('GET', './mock/200', [200, {"Content-Type": "text/html"}, '']);
+                server.respondWith('GET', '/mock/smth', [200, {"Content-Type": "text/html"}, '<div id="smth">smth</div>']);
                 server.autoRespond = true;
```
Comment (warpech) [inline test/juicy-html_no-content.html]: See the comment above about fake server `restore()`

File: test/juicy-html_inline.html [hunk 3]
```diff
@@ -61,9 +63,9 @@
...
 
                 // https://github.com/Juicy/juicy-html/issues/17
-                it('should remove stamped nodes after `content` attribute is removed', function(){
+                it('should remove stamped nodes after `html` attribute is removed', function(){
                     var parent = myEl.parentNode;
-                    myEl.removeAttribute('content');
+                    myEl.removeAttribute('html');
                     expect(parent.children.length).to.be.equal(1);
                     expect(parent.childNodes.length).to.be.equal(1);
                 });
```
Comment (alshakero) [inline test/juicy-html_inline.html]: Same as above. Could you please add a control `expect`.

File: test/juicy-html_skipping.html [hunk 4]
```diff
@@ -24,42 +24,69 @@
...
-                sinon.spy(myEl, 'skipStampingPendingFile');
+            context('when href attribute is changed', function() {
+                it('to another url, skipStampingPendingFile method should get called', function(){
+                    myEl.setAttribute('href', '../examples/partials/page_1.html');
+                    expect(myEl.skipStampingPendingFile).to.be.calledOnce;
+                });
+                it('- removed, skipStampingPendingFile method should get called', function(){
+                    myEl.removeAttribute('href');
+                    expect(myEl.skipStampingPendingFile).to.be.calledOnce;
+                });
```
Comment (alshakero) [inline test/juicy-html_skipping.html]: Could you please add a control `expect` to make sure `myEl.skipStampingPendingFile` was never called before the `removeAttribute`.

File: test/juicy-html_skipping.html [hunk 5]
```diff
@@ -24,42 +24,69 @@
...
-                myEl.setAttribute('content', '../examples/partials/concatenated.html');
-                expect(myEl.skipStampingPendingFile).to.be.calledOnce;
+            context('when href property is changed', function() {
+                it('to another url, skipStampingPendingFile method should get called', function(){
+                    myEl.href = '../examples/partials/page_1.html';
+                    expect(myEl.skipStampingPendingFile).to.be.calledOnce;
+                });
+                it('to `null`, skipStampingPendingFile method should get called', function(){
+                    myEl.href = null;
+                    expect(myEl.skipStampingPendingFile).to.be.calledOnce;
```
Comment (alshakero) [inline test/juicy-html_skipping.html]: Could you please add a control `expect` to make sure `myEl.skipStampingPendingFile` was never called before the `myEl.href`.

**Comment by tomalec** [comment]
As those were changes mostly related to juicy-html part, I made a PR to juicy-html repo. https://github.com/Juicy/juicy-html/pull/32

**Comment by tomalec** [comment]
@alshakero  @warpech  Travis was green, then it failed with README.md update, I'll try to debug travis, but I believe imported-template changes are ready for another review

File: test/content-legacy/inline.html [hunk 6]
```diff
@@ -37,6 +37,10 @@
                 });
 
                 it('should remove stamped nodes after content is cleared', function(){
+                    var parent = myEl.parentNode;
+                    // pre-test conditions
+                    expect(parent.children.length).to.be.equal(2);
+                    // actual test
```
Comment (alshakero) [inline test/content-legacy/inline.html]: I _think_ this pre-test would make more sense if it was of the same nature of the test. Sth like `expect(myEl.nextElementSibling).not.to.be.null`

And I don't really understand this test. 

```js
expect(parent.children.length).to.be.equal(2);
myEl.setAttribute('content', null);
expect(myEl.parentNode.children.length).to.equal(2)
```
So `myEl.setAttribute('content', null);` isn't doing anything? 

_(I apologize if this doesn't belong to this PR)_

File: test/inline.html [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                it('should stamp it into DOM as a sibling', function(){
+                    expect(myEl.nextElementSibling).to.be.not.null;
+                    expect(myEl.nextElementSibling.tagName).to.equal('H1');
+                    expect(myEl.nextElementSibling.innerHTML).to.equal('Hello World');
+                });
+
+                it('should remove stamped nodes after html is cleared', function(){
+                    var parent = myEl.parentNode;
+                    // pre-test conditions
+                    expect(parent.children.length).to.be.equal(2);
```
Comment (alshakero) [inline test/inline.html]: Same as above, IMHO, pre-test only makes sense if it uses the same logic but different values. 
`expect(myEl.nextElementSibling).not.to.be.null` would be much easier to read than `expect(parent.children.length).to.be.equal(2)` and makes more sense.

File: test/inline.html [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                // https://github.com/Juicy/juicy-html/issues/17
+                it('should remove stamped nodes after `html` attribute is removed', function(){
+                    var parent = myEl.parentNode;
+                    // pre-test conditions
+                    expect(parent.children.length).to.be.equal(2);
+                    // actual test
+                    myEl.removeAttribute('html');
+                    expect(parent.children.length).to.be.equal(1);
+                    expect(parent.childNodes.length).to.be.equal(1);
+                });
```
Comment (alshakero) [inline test/inline.html]: Just wanted to compliment this. This looks super neat and robust.

File: test/content-legacy/no-content.html [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+                    expect(warnSpy.withArgs(sinon.match(/empty/))).to.be.calledTwice; // once per element
+                });
+                describe('[not desired but expected]', function(){
+                    it('should call not call <code>console.warn</code> in pollyfiled HTML Imports until webcomponents/webcomponentsjs/issues/535 gets fixed/merged', function () {
+                        if(HTMLImports && HTMLImports.useNative === false){
+                            expect(warnSpy.withArgs(sinon.match.any, sinon.match.any, sinon.match(/empty/))).not.to.be.called;
+                            expect(warnSpy.withArgs(sinon.match.any, sinon.match.any, sinon.match(/empty/))).not.to.be.calledTwice; // once per element
+                        }
+                    });
+                    it('should call call <code>console.warn</code> for native HTML Imports ', function () {
```
Comment (alshakero) [inline test/content-legacy/no-content.html]: `call` is repeated.

File: test/content-legacy/no-content.html [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+                it('should remove previously stamped nodes', function () {
+                    expect(changedEl.nextElementSibling).to.be.null;
+                    expect(changedEl.previousElementSibling).to.be.null;
+                });
+                xit('should call <code>console.warn</code> with message about "empty"', function () {
+                    expect(warnSpy.withArgs(sinon.match(/empty/))).to.be.called;
+                    expect(warnSpy.withArgs(sinon.match(/empty/))).to.be.calledTwice; // once per element
+                });
+                describe('[not desired but expected]', function(){
+                    it('should call not call <code>console.warn</code> in pollyfiled HTML Imports until webcomponents/webcomponentsjs/issues/535 gets fixed/merged', function () {
```
Comment (alshakero) [inline test/content-legacy/no-content.html]: typo in `it` description.

File: test/content-legacy/inline.html (hunk 6 above)
Comment (tomalec) [reply test/content-legacy/inline.html]: Where you have found:
```js
expect(parent.children.length).to.be.equal(2);
myEl.setAttribute('content', null);
expect(myEl.parentNode.children.length).to.equal(2)
```

This is 
```
                    // pre-test conditions
                    expect(parent.children.length).to.be.equal(2);
                    // actual test
                    myEl.setAttribute('content', null);
                    expect(myEl.nextElementSibling).to.be.null;
                    expect(myEl.previousElementSibling).to.be.null;
```

`expect(myEl.nextElementSibling).not.to.be.null` would not test the thing we are trying to test:
"should remove stamped nodes" <=> there are no nodes stamped where they were  <=> `juicy-html` is the only sibling

there were stamped nodes <=> there was `juicy-html` and something <=> `juicy-html` had previous OR next sibling

You would have to test `expect(myEl.nextElementSibling).not.to.be.null or expect(myEl.prevousElementSibling).not.to.be.null` which is not doable with mocha.

File: test/inline.html (hunk 7 above)
Comment (tomalec) [reply test/inline.html]: See answer above. 

Please then write full requirements of "pre-test" as I start to think we are going to: write code, then write tests for code, then write the tests for tests of code, then...

We already have 
```html
    <test-fixture id="juicy-html-fixture">
        <template>
            <!-- nest to workaround test-fixture bug -->
            <div><template is="imported-template" html="<h1>Hello World</h1>">
            </template></div>
        </template>
    </test-fixture>
```
```js
beforeEach(function(done) {
     myEl = fixture('juicy-html-fixture').querySelector('template[is="imported-template"]');
```

If you are afraid that `html` attribute may not work and we may not have any sibling in place before we remove the `html` attribute, we have a test for it already. That;s what the atomicity of tests is for, we are not writing one test to rule them all, we are writing many tests on per each behavior.

If you rely on not having sibling after you cleared `html` attr and this tests passes you should still be fine even if there were nothing before.

If you rely on having some siblings once `html` is set we have a separate tests for it:
 - https://github.com/Juicy/imported-template/blob/bf11ee837cd6070e2ced1600cac3518e8d57c5b3/test/inline.html#L35
 - https://github.com/Juicy/imported-template/blob/bf11ee837cd6070e2ced1600cac3518e8d57c5b3/test/inline.html#L51

File: test/content-legacy/inline.html (hunk 6 above)
Comment (tomalec) [reply test/content-legacy/inline.html]: I could change it to 
```js
                    var parent = myEl.parentNode;
                    // pre-test conditions
                    expect(parent.children.length).to.be.equal(2);
                    // actual test
                    myEl.setAttribute('html', null);
                    expect(parent.children.length).to.be.equal(1);
```
Checking siblings was just IMO equivalent to this, just withour a need to have a parent reference (which was added just for the pre-test)

**Comment by tomalec** [comment]
Yeah, I'd say most of this again belongs to `juicy-html` https://github.com/Juicy/juicy-html/pull/32 I would really like to push property setters to starcounter-include -> Starcounter -> move on with Polymer  and WCv1 upgrade without spending more days on fiddling with existing tests for `juicy-html`

File: test/content-legacy/inline.html (hunk 6 above)
Comment (alshakero) [reply test/content-legacy/inline.html]: Regarding the first snippet, you're right. It's my fault. This fooled me into thinking it was the same test:

image

Didn't see the light-blue strip at all.

Regarding the rest, you're right, and it's just subjectively easier to read, other than that, LGTM.
